### PR TITLE
Use model classes instead of CIMCORE json to construct components.

### PR DESCRIPTION
**Pull request by mint-thompson**
When making constraints and elements to put into the modeldoc, use the model classes from shr-models
in their original forms. This results in some changes to attribute paths.
When an element definition contains a constraint on a field's subfield, and that field is declared by this element (not inherited),
show those constraints in the "Properties Constrained" portion of the model doc.

This pull request is intended to be applied along with accompanying pull requests on shr-cli and shr-models:
* https://github.com/standardhealth/shr-models/pull/48
* https://github.com/standardhealth/shr-cli/pull/226

### CODE REVIEW

File: components/constraints.js [hunk 1]
```diff
@@ -10,9 +10,9 @@ function cardToString(card) {
   let min = 0;
   let max = '*';
   if (card) {
-    if ('min' in card)
+    if (typeof card.min != 'undefined')
```
Comment (cmoesel) [inline components/constraints.js]: In cases like this, I generally prefer
```js
if (card.min != null)
```
It's shorter and easier to read, but it also tests against both `null` and `undefined` -- which is what I think we intended in this case.  You'll still find the `typeof x === 'undefined'` construct in SHR code here or there, but usually that's from _before_ I realized the benefit of `== null`/`!= null`.  (Note -- double _not triple_ equals).

Comment (cmoesel) [reply components/constraints.js]: This shows up below as well, but I won't call it out again.

File: components/constraints.js [hunk 2]
```diff
@@ -96,139 +96,115 @@ class Constraints {
...
   includesCode(constraint, subpath) {
-    constraint.forEach((item) => {
-      const system = item.system ? item.system : '';
-      const name = 'Includes Code';
-      const value = `${system} (code: ${item.code})`;
-      const lastMod = item.lastModifiedBy;
-      const iConstraint = this.newConstraint(name, value, subpath, lastMod);
-      this.constraints.push(iConstraint);
-    });
+    const system = constraint.code.system ? constraint.code.system : '';
```
Comment (cmoesel) [inline components/constraints.js]: The above is another construct you probably saw in other SHR code (probably written by me!), but I think can be simplified to:
```js
const system = constraint.code.system || '';
```

File: components/constraints.js [hunk 3]
```diff
@@ -96,139 +96,115 @@ class Constraints {
...
-      break;
-    case 'card':
-      this.cardConstraint(constraint, subpath);
-      break;
-    default:
-      logger.warn('Unknown constraint type %s. ERROR_CODE:TBD', cType);
+  buildFullPath(constraint) {
+    if(constraint.path.length > 0) {
+      return constraint.path.reduce((fullPath, pathPiece) => {
+        if(pathPiece.fqn in this.elements) {
```
Comment (cmoesel) [inline components/constraints.js]: Are primitives in `this.elements`?  I feel like they're not, but I could be wrong (I don't have much experience in this module).  If primitives are not in `this.elements` then are we dropping them when building the full path like this?

File: components/constraints.js [hunk 4]
```diff
@@ -239,18 +215,42 @@ class Constraints {
...
-      if (modConstraint) {
-        this.switchConstraintType(modConstraint, 'card', this.field.name);
-      }
     }
     // Now look at the rest of the constraints
-    if ('constraints' in this.field) {
-      Object.keys(this.field.constraints).forEach((cType) => {
-        const constraint = this.field.constraints[cType];
-        this.switchConstraintType(constraint, cType, this.field.name);
+    if ('constraints' in this.field && 'constraintsFilter' in this.field) {
```
Comment (cmoesel) [inline components/constraints.js]: Trivial thing, but I always thought it was kind of weird the way Abhijay did this.  Maybe we should go ahead and use the more traditional:
```js
if (this.field.constraints && this.field.constraintsFilter) {
```

**Review by cmoesel — CHANGES_REQUESTED** [review]
The code looks really good -- and kudos for your attention to all the little details in making this work (there are a lot!).  I put a few comments in on minor adjustments to consider.

I tried to run this, but was unable to.  Here is my branch setup:
```
shr-cli                  [remove-cimcore-export]
shr-data-dict-export     [master]
shr-es6-export           [master]
shr-expand               [master]
shr-fhir-export          [master]
shr-grammar              [master]
shr-json-javadoc         [remove-cimcore-export]
shr-json-schema-export   [master]
shr-models               [remove-cimcore-export]
shr-test-helpers         [master]
shr-text-import          [master]
shr_spec                 [master]
```
I ran `./yarn-link-all` and then executed the CLI and got this error:
```
/Users/cmoesel/dev/standardhealth/shr-cli/app.js:93
const errorFiles = [shrTI.errorFilePath(), shrEx.errorFilePath(), shrFE.errorFilePath(), shrJDE.errorFilePath(),
                                                                                                ^

TypeError: shrJDE.errorFilePath is not a function
    at Object.<anonymous> (/Users/cmoesel/dev/standardhealth/shr-cli/app.js:93:97)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:643:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:839:10)
    at internal/main/run_main_module.js:17:11
```

I think that Nick's new logger refactoring snuck in on master under you and you need to rebase to pull in his changes.  Once you've done that, let me know and I'll give it a try again.

Feel free to reach out if you have any questions!

File: components/constraints.js (hunk 3 above)
Comment (mint-thompson) [reply components/constraints.js]: `this.elements` does not contain primitives, so you are correct in identifying that this code results in the primitive's name not being part of the path. This is consistent with the module's previous behavior.
Given that the intent is to avoid adding the primitive's name to the path, this intent is clearer by directly checking `pathPiece.namespace == 'primitive'`. Does that sound reasonable to you?

**Review by cmoesel — APPROVED** [review]
Nice job, @mint-thompson.  Code looks good and modeldoc looks right.  Did a diff on modeldoc outputs and selected random element pages -- and all diffs made sense.  Couldn't do it for every single one (lots of whitespace-only changes muddying the waters), but based on random spot-checks, looks great!
